File: harnesses/cando/driver.template.c
/* cando-style function-level differential driver — C ORACLE side.
 *
 * Build this against the C library under port; the Rust driver
 * (driver.template.rs) is the same protocol built against the Rust port. The
 * harness (cando_diff.py) runs both over a vector suite and diffs, so each
 * exported function is tested individually — the differential a *library* needs,
 * where diff_run.py (whole-program argv->stdout) does not fit.
 *
 * Protocol:
 *   argv[1] = function name, argv[2..] = string args, optional stdin.
 *   Print the function's result CANONICALLY to stdout — the SAME textual format
 *   the Rust driver prints (fix byte order, radix, separators, trailing newline
 *   on both sides, or you manufacture a divergence). Return 0 on success,
 *   nonzero on error / unsupported function. A nonzero exit makes cando_diff
 *   treat the vector as a failed C baseline (a vector must pass on C before it
 *   can judge Rust) unless the suite is run with --allow-oracle-error.
 *
 * Keep the driver THIN: parse args, call ONE library function, serialize the
 * result. All the logic under test lives in the library, not here.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/* #include "yourlib.h" */

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <func> [args...]\n", argv[0]);
        return 2;
    }
    const char *func = argv[1];

    /* One branch per exported function under test. Replace with the real calls. */
    if (strcmp(func, "__EXAMPLE__checksum") == 0) {
        /* unsigned r = yourlib_checksum((const unsigned char *)argv[2], strlen(argv[2])); */
        /* printf("%u\n", r); */
        return 0;
    }
    if (strcmp(func, "__EXAMPLE__parse") == 0) {
        /* int rc = yourlib_parse(argv[2]); */
        /* printf("%d\n", rc); */
        return 0;
    }

    fprintf(stderr, "unsupported function: %s\n", func);
    return 4;
}

File: harnesses/cando/driver.template.rs
//! cando-style function-level differential driver — RUST side.
//!
//! Build this against the Rust port; it is the same protocol as
//! driver.template.c (the C oracle side). The harness (cando_diff.py) runs both
//! over a vector suite and diffs, testing each ported function individually.
//!
//! Protocol:
//!   argv[1] = function name, argv[2..] = string args, optional stdin.
//!   Print the result CANONICALLY to stdout — byte-identical format to the C
//!   driver (same radix, separators, trailing newline). Exit 0 on success,
//!   nonzero on error / unsupported function.
//!
//! Keep the driver THIN: parse args, call ONE library function, serialize the
//! result. The logic under test lives in the crate, not here. No `unwrap()` on
//! the argv/stdin the harness feeds you — return a nonzero exit instead, so a
//! bad vector is a clean baseline failure, not a panic.

use std::process::exit;

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let Some(func) = args.first().map(String::as_str) else {
        eprintln!("usage: driver <func> [args...]");
        exit(2);
    };

    match func {
        // One arm per exported function under test. Replace with the real calls.
        "__EXAMPLE__checksum" => {
            // let r = yourcrate::checksum(args[1].as_bytes());
            // println!("{r}");
        }
        "__EXAMPLE__parse" => {
            // match yourcrate::parse(&args[1]) {
            //     Ok(v) => println!("{v}"),
            //     Err(_) => exit(1),
            // }
        }
        other => {
            eprintln!("unsupported function: {other}");
            exit(4);
        }
    }
}
